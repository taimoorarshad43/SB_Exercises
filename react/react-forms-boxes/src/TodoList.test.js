import React from "react";
import { render, fireEvent, queryAllByText } from "@testing-library/react";
import TodoList from "./TodoList";

it("renders without crashing", function() {
  render(<TodoList />);
});

it("matches snapshot", function() {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("adds a new todo when form is submitted", function() {
  const { getByLabelText, getByText, queryByText } = render(<TodoList />);
  
  const taskInput = getByLabelText("Task:");
  const submitButton = getByText("Add a todo!");
  
  fireEvent.change(taskInput, { target: { value: "Test task" } });
  fireEvent.click(submitButton);
  
  // Check that a todo with "X" button was added
  const removeButtons = queryByText("X");
  expect(removeButtons).toBeInTheDocument();
  expect(getByText("Test task")).toBeInTheDocument();
});

it("removes a todo when X button is clicked", function() {
  const { getByLabelText, getByText, queryByText } = render(<TodoList />);
  
  // Add a todo first
  const taskInput = getByLabelText("Task:");
  const submitButton = getByText("Add a todo!");
  
  fireEvent.change(taskInput, { target: { value: "Test task" } });
  fireEvent.click(submitButton);
  
  // Verify todo was added
  let removeButtons = queryByText("X");
  expect(removeButtons).toBeInTheDocument();
  expect(getByText("Test task")).toBeInTheDocument();
  
  // Remove the todo
  fireEvent.click(removeButtons);
  
  // Verify todo was removed
  removeButtons = queryByText("X");
  expect(removeButtons).not.toBeInTheDocument();
  expect(queryByText("Test task")).not.toBeInTheDocument();
});

it("can add and remove todos", function() {
  const { getByLabelText, getByText, queryByText, container } = render(<TodoList />);
  
  // Add multiple todos
  const taskInput = getByLabelText("Task:");
  const submitButton = getByText("Add a todo!");
  
  fireEvent.change(taskInput, { target: { value: "First task" } });
  fireEvent.click(submitButton);
  
  fireEvent.change(taskInput, { target: { value: "Second task" } });
  fireEvent.click(submitButton);
  
  // Verify both todos were added
  expect(getByText("First task")).toBeInTheDocument();
  expect(getByText("Second task")).toBeInTheDocument();
  
  // Remove first todo
  const removeButtons = queryAllByText(container, "X");
  fireEvent.click(removeButtons[0]);
  
  // Verify first todo was removed
  expect(queryByText("First task")).not.toBeInTheDocument();
  expect(getByText("Second task")).toBeInTheDocument();
});
