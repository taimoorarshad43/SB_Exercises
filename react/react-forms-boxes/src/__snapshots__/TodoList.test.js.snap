// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Todo App
    </h1>
    <div
      style="margin-bottom: 20px; padding: 20px; border: 1px solid #ccc; border-radius: 5px;"
    >
      <h3>
        Add a New Todo
      </h3>
      <form>
        <label
          for="task"
        >
          Task: 
        </label>
        <input
          id="task"
          name="task"
          placeholder="Enter a new task..."
          style="margin-left: 5px; padding: 8px; width: 300px; margin-right: 10px;"
          type="text"
          value=""
        />
        <button
          style="padding: 8px 16px; background-color: rgb(0, 123, 255); color: white; border-radius: 3px; cursor: pointer;"
          type="submit"
        >
          Add a todo!
        </button>
      </form>
    </div>
    <ul
      style="list-style: none; padding: 0px;"
    />
  </div>
</DocumentFragment>
`;
